Periodic IPTables sync

siddhant94 opened the pull request:
Fixes #628 . 
Add a long running goroutine which periodically syncs antrea required iptables rules on linux.
If the sync call/operation fails (for example - xtables lock contention), the next attempt would occur at the next sync interval.

antoninbas commented on test/integration/agent/route_test.go:
```diff
@@ -232,6 +231,23 @@ func TestInitialize(t *testing.T) {
 			assert.Equal(t, expectedData, string(actualData), "mismatch iptables data in table %s", table)
 		}
 	}
+	t.Log("Testing ipTables syncLoop")
```
I feel like this should be a separate test case

siddhant94 replied:
Added a new test case.

antoninbas commented on pkg/agent/route/route_linux.go:
```diff
@@ -115,19 +116,46 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
 func (c *Client) initIPTablesOnce(done func()) {
```
maybe the name of this function is not appropriate any more?

siddhant94 replied:
Removed name, moved to anonymous fn. Same behaviour.

antoninbas commented on pkg/agent/route/route_linux.go:
```diff
@@ -115,19 +116,46 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
 func (c *Client) initIPTablesOnce(done func()) {
 	defer done()
 	backoffTime := 2 * time.Second
+	stopCh := signals.RegisterSignalHandlers()
```
We do not call this function multiple times. We call it once in `main()` and then pass the `stopCh` channel around to all components as needed. IMO, calling `RegisterSignalHandlers` here like you are doing makes it hard to unit test this function down the line if needed.

siddhant94 replied:
Moved start of goroutine to agent.go ([as per feedback](https://github.com/vmware-tanzu/antrea/pull/1751#discussion_r558579670)). So now we have access to the stopCh, hence removed call to RegisterSignalHandlers.

antoninbas commented on pkg/agent/route/route_linux.go:
```diff
@@ -115,19 +116,46 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+	c.syncIPTablesWithRetry(backoffTime)
+	klog.Info("Successfully synced iptables")
+	klog.Info("Starting sync loop for IPTables")
+	go c.ipTablesSyncLoop(stopCh)
```
We try to start all long-lived goroutines from https://github.com/vmware-tanzu/antrea/blob/master/cmd/antrea-agent/agent.go, please try to follow the same pattern.

siddhant94 replied:
Updated the PR.

antoninbas commented on pkg/agent/route/route_linux.go:
```diff
@@ -115,19 +116,46 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+	var syncPeriod = 30 * time.Second
+	var syncTicker = time.NewTicker(syncPeriod)
+	defer syncTicker.Stop()
+	for {
```
We typically use K8s utility functions like https://godoc.org/k8s.io/apimachinery/pkg/util/wait#Until for these scenarios

siddhant94 replied:
Removed go's ticker with wait.Until. Updated the PR.

antoninbas commented on pkg/agent/route/route_linux.go:
```diff
@@ -115,19 +116,46 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+				klog.Errorf("Failed to sync ipset: %v - will retry in %v", err, syncPeriod)
+				continue
+			}
+			c.syncIPTablesWithRetry(5 * time.Second)
```
Pleas make the sync period a top-level named `const`

siddhant94 replied:
Moved this to top level constant in agent.go, so that we can control the sync interval. For ex- integration tests uses a much smalled interval to speed up it's sync test. Currently, I have kept sync interval at 60 sec. Let me know what value do you suggest?

antoninbas commented on test/integration/agent/route_test.go:
```diff
@@ -232,6 +231,23 @@ func TestInitialize(t *testing.T) {
+	deleteRuleCmd := deleteOption + strings.Join([]string{"-m", "comment", "--comment", "\"Antrea: jump to Antrea prerouting rules\"", "-j", "ANTREA-PREROUTING"}, " ")
+
+	// #nosec G204: ignore in test code
+	actualData, err := exec.Command("bash", "-c", deleteRuleCmd).Output()
```
I notice other tests doing the same thing (calling `iptables` through `bash`), but I wonder why the `iptables` executable cannot be invoked directly instead?

antoninbas reviewed (commented):
Thanks for opening this PR, I left some comments. Please make sure you sign the CLA as well.

siddhant94 commented:
Hi @antoninbas , as per feedback,
- Use wait.Until
- Start long running from agent.go
- Pass stopCh, called only once
- Make sync interval top level constant
- Separate integration test

I have addressed these, let me know if implementation is still inconsistent.  
Also needed feedback on couple of things -
1) sync interval is set at 60sec, and,  
2) The sync behavior is such that we sync/restore iptables, but if during that, we encounter an error, the sync fn returns, thus next attempt to sync would effectively be the next iteration of sync i.e. after sync interval.

siddhant94 commented:
Looking into ci failures

siddhant94 commented:
> Looking into ci failures

Fixed, updated the PR.

tnqn commented on pkg/agent/route/route_linux.go:
```diff
@@ -110,24 +116,30 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+// Run waits for node network initialization, then periodically syncs the iptables rules.
+// It will not return until stopCh is closed.
+func (c *Client) Run(stopCh, networkReadyCh <-chan struct{}, syncInterval time.Duration) {
+	<-networkReadyCh
```
I think this routine doesn't really have dependency on the readiess of the whole network, we just need to avoid it run concurrently with syncIPSet and syncIPTables, right? Maybe use an internal variable/channel to check if initialization is done to avoid unnecessary external dependency.
syncInternal could be this file internal too since it's not shared or customized on caller side.

siddhant94 replied:
So the reason for blocking on `networkReadyCh` is to get signalled when initialisation is done. It closes networkReadyCh as soon as we are able to init iptables. I'll paste the snippet and file and method names below,

`agentInitializer.Initialize()` method starts a goroutine which waits on `routeClient.Initialize()` to signal all synchronous and asynchronous initializations are done. Asynchronous being the init of iptables.
```
// file: pkg/agent/agent.go
wg.Add(1)
if err := i.routeClient.Initialize(i.nodeConfig, wg.Done); err != nil {
	return err
}

// The Node's network is ready only when both synchronous and asynchronous initialization are done.
go func() {
	wg.Wait()
	close(i.networkReadyCh)
}()
```
The done() fn is only called when init iptables is done by routeClient.

tnqn replied:
Thanks for the explanation. I understand blocking on networkReadyCh means all sync and async initialization done.   What I meant was the routine that resyncs iptables/ipset doesn't need to wait for that. The signal was designed for CNIServer which requires it, and the routeClient was made agnostic to the existence of networkReadyCh by using a callback function by purpose.

Passing the networkReadyCh here actually makes `Run` wait for `syncIPTablesWithRetry` to be done but via an external event, which looks unnecessary to me.

I was suggesting using an internal bool to sync the two routines, for instance, `syncIPTablesWithRetry` sets initialized to true and `Run` waits for it to be true.

siddhant94 replied:
Got it, my bad. It makes sense to signal init internally. I'll make the changes.

tnqn commented on pkg/agent/route/route_linux.go:
```diff
@@ -94,13 +95,18 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+	go func() {
+		defer done()
+		var backoffTime = 2 * time.Second
+		c.syncIPTablesWithRetry(backoffTime)
```
The routine that initializes the iptables once is simple, I think we either keep all of its logic in one function as it was or keep them in one anonymous without extracting part of it to another function, given that non function is reused or unit tested.

siddhant94 replied:
Sure @tnqn , I think keeping it anonymous should do, and I'll keep the the xtables contention issue highlighted via comments, as to the reason we want to retry while initialisation. Should need be later, we can extract this into a named method.

tnqn replied:
I think L103 already has the reason why retry is needed.

tnqn commented on pkg/agent/route/route_linux.go:
```diff
@@ -110,24 +116,30 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+		klog.Errorf("failed to sync iptables: %v", err)
 		return
 	}
+	klog.V(3).Infof("Successfully synced node IpTables.")
```
```suggestion
	klog.V(3).Infof("Successfully synced node iptables")
```

tnqn reviewed (commented):
@siddhant94 thanks for the PR, I left some comments.

siddhant94 commented:
> @siddhant94 thanks for the PR, I left some comments.

Updated the PR @tnqn

antoninbas commented on pkg/agent/route/route_linux.go:
```diff
@@ -62,6 +63,8 @@ var (
 	// globalVMAC is used in the IPv6 neighbor configuration to advertise ND solicitation for the IPv6 address of the
 	// host gateway interface on other Nodes.
 	globalVMAC, _ = net.ParseMAC("aa:bb:cc:dd:ee:ff")
+	// iptablesInited is used to determine when iptables initialisation is done.
```
s/initialisation/initialization

I'm not sure "inited" is an actual word, can we rename the variable `iptablesInitialized`?

siddhant94 replied:
definitely, I was just wary that it might be too long.
Updated

antoninbas commented on pkg/agent/route/route_linux.go:
```diff
@@ -94,13 +97,26 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+		var backoffTime = 2 * time.Second
+		for {
+			if err := c.syncIPTables(); err != nil {
+				klog.Errorf("failed to initialize iptables: %v - will retry in %v", err, backoffTime)
```
we do not capitalize error messages but we capitalize logs: s/failed to.../Failed to...

siddhant94 replied:
updated

antoninbas commented on pkg/agent/route/route_linux.go:
```diff
@@ -110,24 +126,34 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+// Run waits for iptables initialization, then periodically syncs iptable rules.
+// It will not return until stopCh is closed.
+func (c *Client) Run(stopCh <-chan struct{}, syncInterval time.Duration) {
 	for {
```
is that a busy loop? why don't we use a channel instead?

siddhant94 replied:
Ya we should 👍🏼 , updated

antoninbas commented on test/integration/agent/route_test.go:
```diff
@@ -234,6 +233,51 @@ func TestInitialize(t *testing.T) {
+		assert.Equal(t, "", string(actualData), "failed to remove iptables rule for %v", tc)
+	}
+	stopCh := make(chan struct{})
+	syncInterval := 15 * time.Second
+	go routeClient.Run(stopCh, syncInterval)
```
could we use a smaller interval for testing?

siddhant94 replied:
Made it 2sec

tnqn reviewed (commented):
Thanks @siddhant94.  One comment about the interface argument, otherwise LGTM.

siddhant94 commented on cmd/antrea-agent/agent.go:
```diff
@@ -253,6 +255,8 @@ func run(o *Options) error {
 
 	log.StartLogFileNumberMonitor(stopCh)
 
+	go routeClient.Run(stopCh, ipTablesSyncInterval)
```
Agreed @tnqn , this should also stay internal. I missed this in previous iteration. I was thinking of exposing a function `SetSyncInterval` which would override the variable, it would be meant to be called from route package's Initialization() method and from the integration test.
Let me know if this is inconsistent. Also, if there's someplace in antrea code where we have a similar issue, I can look at that and change this part.

tnqn replied:
I forgot that the integration test is in a different package. I'm ok with just declaring the variable as public, but we should comment why it's made so. We had a similar case before that an integration test needs to use package internal consistant/variable and it just made them public to solve it. However, that code has been deleted when implmenetation evolves.

@antoninbas what do you think?

antoninbas replied:
typically my favorite approach is to have a way to pass it to the constructor:

```go
const ipTablesSyncInterval = 60 * time.Second

type ClientConfig struct {
    IPTablesSyncInterval time.Duration
}

func NewClient(serviceCIDR *net.IPNet, networkConfig *config.NetworkConfig, noSNAT bool, configFns ...func(*ReporterConfig)) *Client {
    config := ClientConfig{
        IPTablesSyncInterval: defaultIPTablesSyncInterval,
    }
    for _, fn := range configFns {
        fn(&config)
    }
    return &Client{
        ClientConfig: config,
        // other members
    }
}
```
This way the call to `NewClient` does not need to change in `cmd/antrea-agent/agent.go`, but for the integration tests (or unit tests for that matter), we can easily provide a different value for the sync interval.

Of course that may be a bit overkill here with a single configuration parameter. So if you prefer to make the variable public and add a comment, that's fine by me.

siddhant94 replied:
@antoninbas I actually missed this last comment, I just noticed it after push.

tnqn commented on pkg/agent/route/route_linux.go:
```diff
@@ -110,24 +130,30 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+func (c *Client) syncIPInfra() {
+	// Sync ipset before syncing iptables rules
+	if err := c.syncIPSet(); err != nil {
+		klog.Errorf("failed to sync ipset: %v", err)
```
```suggestion
		klog.Errorf("Failed to sync ipset: %v", err)
```

tnqn commented on pkg/agent/route/route_linux.go:
```diff
@@ -110,24 +130,30 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig, done func()) error {
+		return
+	}
+	if err := c.syncIPTables(); err != nil {
+		klog.Errorf("failed to sync iptables: %v", err)
```
```suggestion
		klog.Errorf("Failed to sync iptables: %v", err)
```

tnqn commented on pkg/agent/route/route_windows.go:
```diff
@@ -152,6 +152,11 @@ func (c *Client) UnMigrateRoutesFromGw(route *net.IPNet, linkName string) error
 	return errors.New("UnMigrateRoutesFromGw is unsupported on Windows")
 }
 
+// Run is not supported on Windows and returns immediately.(On linux, it syncs iptables periodically)
```
No need to say what it does on other platform

tnqn commented on pkg/agent/route/interfaces.go:
```diff
@@ -44,4 +44,7 @@ type Interface interface {
 	// if linkName is nil, it should remove the routes.
 	UnMigrateRoutesFromGw(route *net.IPNet, linkName string) error
+
+	// Run periodically syncs iptables.
```
The interface is for windows too, could we avoid using linux specific concept like other methods.

siddhant94 replied:
Updated to - "Run periodically syncs firewall rules required for routing in host network."

tnqn reviewed (commented):
@siddhant94 The commit message seems having a very long title or missing empty line between title and body. Could you update it following https://github.com/golang/go/wiki/CommitMessage?

siddhant94 commented:
> @siddhant94 The commit message seems having a very long title or missing empty line between title and body. Could you update it following https://github.com/golang/go/wiki/CommitMessage?

Updated it as per the wiki.

tnqn commented on pkg/agent/route/interfaces.go:
```diff
@@ -44,4 +44,7 @@ type Interface interface {
 	// if linkName is nil, it should remove the routes.
 	UnMigrateRoutesFromGw(route *net.IPNet, linkName string) error
+
+	// Run periodically syncs required firewall rules.
```
firewall rules are not very suitable for what it actually does.
Maybe we don't need to be specific and could just say "Run starts the sync loop."

tnqn reviewed (commented):
@siddhant94 The body of commit message should be wrapped to ~76 characters, otherwise it wouldn't look good from `git log`. And I think you don't need to add "pkg/agent/route:" as prefix as the other words can already scope the change well.

siddhant94 commented:
> @siddhant94 The body of commit message should be wrapped to ~76 characters, otherwise it wouldn't look good from `git log`. And I think you don't need to add "pkg/agent/route:" as prefix as the other words can already scope the change well.

@tnqn I updated the Run fn comment and removed the said prefix from commit, but, for commit body, I could only reduce it to ~100 chars. Should I just omit the body? Just keep the "Fixes" line and title/subject line.

tnqn commented:
@siddhant94 By wrapping the body of commit message, I meant to limit each line ~76 chars.
For example, this is how the current git log shows:
```
commit 5cbe43181f6143f13b683f4a589dde30dfb492c9
Author: Siddhant Sinha <sid@pop-os.localdomain>
Date:   Thu Jan 14 15:12:55 2021 +0530

    Periodically sync antrea required iptables rules

    Add a long-running goroutine, which waits for iptables to be initialised, before periodically syncing it.

    Fixes #628
```

By wrapping it, it appears as:
```
commit 5cbe43181f6143f13b683f4a589dde30dfb492c9
Author: Siddhant Sinha <sid@pop-os.localdomain>
Date:   Thu Jan 14 15:12:55 2021 +0530

    Periodically sync antrea required iptables rules

    Add a long-running goroutine, which waits for iptables to be
    initialised, before periodically syncing it.

    Fixes #628

```

siddhant94 commented:
@tnqn I missed the "wrapped" word, thus misunderstood it completely. I have updated the commit msg, with each line within ~76 chars.

tnqn reviewed (approved):
LGTM, thanks @siddhant94

siddhant94 commented:
> LGTM, thanks @siddhant94

Thanks @tnqn @antoninbas for the patient iterations. Learnt new stuff and got to know a lot of antrea context also.

tnqn commented:
/test-all

tnqn commented:
/test-windows-conformance
/test-windows-networkpolicy

tnqn commented:
/test-windows-conformance
/test-windows-networkpolicy

tnqn commented:
@lzhecheng do you know why the commands cannot trigger windows e2e tests?

lzhecheng commented:
> @lzhecheng do you know why the commands cannot trigger windows e2e tests?

There was something wrong with Jenkins smee service. Now it should work.
/test-windows-conformance
/test-windows-networkpolicy

tnqn commented:
/test-windows-conformance
/test-windows-networkpolicy

lzhecheng commented:
`windows-conformance` is pending now but it has actually succeeded, checked on Jenkins side.
